### ISSUE TITLE
Apps Open Files (app.files) reports more than 100% usage

### ISSUE THREAD

**Issue by arnowelzel** [netdata/netdata#10327]
##### Bug report summary

I get alerts by Apps Open Files (app.files) that the number of open files is 177%. However the total number is way less than limit of 100000. It looks like the calculation is done based on a limit of 1024 and the not the real limit.

![grafik](https://user-images.githubusercontent.com/6613614/101241113-b6fe3d00-36f3-11eb-9c92-33be79e53011.png)

![grafik](https://user-images.githubusercontent.com/6613614/101241130-ddbc7380-36f3-11eb-9f08-88212713e9af.png)

```
# cat /proc/sys/fs/file-max
100000
```

##### OS / Environment

```
Linux example.com 4.15.0-126-generic #129-Ubuntu SMP Mon Nov 23 18:53:38 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=18.04
/etc/lsb-release:DISTRIB_CODENAME=bionic
/etc/lsb-release:DISTRIB_DESCRIPTION="Ubuntu 18.04.5 LTS"
/etc/os-release:NAME="Ubuntu"
/etc/os-release:VERSION="18.04.5 LTS (Bionic Beaver)"
/etc/os-release:ID=ubuntu
/etc/os-release:ID_LIKE=debian
/etc/os-release:PRETTY_NAME="Ubuntu 18.04.5 LTS"
/etc/os-release:VERSION_ID="18.04"
/etc/os-release:HOME_URL="https://www.ubuntu.com/"
/etc/os-release:SUPPORT_URL="https://help.ubuntu.com/"
/etc/os-release:BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
/etc/os-release:PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
/etc/os-release:VERSION_CODENAME=bionic
/etc/os-release:UBUNTU_CODENAME=bionic
```

##### Netdata version

netdata v1.26.0-317-nightly

##### Component Name

apps.plugin

##### Steps To Reproduce

1. Reboot the machine
2. Wait until number of total open files will raise above 1024
3. See the alert by Apps Open Files (app.files) that the percentage is more than 100%

##### Expected behavior

The percentage is calculated correctly based on the real limit of 100000 and not 1024.

**Comment by ilyam8** [comment]
Hi @arnowelzel 

The alarm and `fdperc` chart variable were added in https://github.com/netdata/netdata/pull/10192

- `/proc/sys/fs/file-max` is per system limit
- [`_SC_OPEN_MAX`](https://github.com/Ancairon/netdata/blob/2ee5379c434c7745811c5da93ef4e663a348e5f6/collectors/apps.plugin/apps_plugin.c#L3618) is [per process limit](https://man7.org/linux/man-pages/man3/sysconf.3.html).

if i understand #10192 correctly it finds max open fds accross all processes (groups in facts, probably this is the problem) and compares it with `_SC_OPEN_MAX`.

---

But yes, something is wrong. @vlvkobal @Ancairon

**Comment by arnowelzel** [comment]
Thanks for the explanation. However - the per process limit on the affected machine is configured to be 4096 and not 1024.

Some code in C to check this:

```
#include <unistd.h>
#include <stdio.h>

int main()
{
        printf("sysconf(_SC_OPEN_MAX) reports: %ld\n", sysconf(_SC_OPEN_MAX));
        return 0;
}
```

And the result:

```
sysconf(_SC_OPEN_MAX) reports: 4096
```

The process with the largest number of file descriptors on the affected machine is MariaDB with about 1787. But this is well within the limit of 4096 and not 176%. If `sysconf(_SC_OPEN_MAX)` does not work properly within netdata it may be a good idea to have a configuration option to override this value.

**Comment by vlvkobal** [comment]
Well, I've checked `_SC_OPEN_MAX` and it depends on `ulimit -n`, so the value can be different for different users. Is there a system-wide limit per process? How can we get it? @Ferroin, any thoughts?

**Comment by vlvkobal** [comment]
We can read the soft limit from `/proc/<pid>/limits` for every process and calculate the alarm based on this data. But I saw a process with the limit = 2. If that process has 2 open files, the fd usage will always be 100% which is not useful.

**Comment by Thumpxr** [comment]
I've got the same problem on my machine aswell.
`cat /proc/sys/fs/file-max
65535`

Usage is at 187%

**Comment by ilyam8** [comment]
> Usage is at 187%

not that bad, my friend showed me an alarm with 1480% 😄 Will wil fix it!

**Comment by ilyam8** [comment]
@vlvkobal we can set limits as variables and calculate percentage in the alarm. So we can add some conditions, for example if limit > N, etc.

**Comment by arnowelzel** [comment]
> 
> 
> @vlvkobal we can set limits as variables and calculate percentage in the alarm. So we can add some conditions, for example if limit > N, etc.

As a default I would set this variable to 1024 - this is the default on Linux.

**Comment by Ancairon** [comment]
> > @vlvkobal we can set limits as variables and calculate percentage in the alarm. So we can add some conditions, for example if limit > N, etc.
> 
> As a default I would set this variable to 1024 - this is the default on Linux.

This was my initial way of doing this, I had a function that was reading from a file the limits and initializing as you said to 1024 (although I was pointing to file-max which was wrong). But from the suggestion https://github.com/netdata/netdata/pull/10192#issuecomment-731748674 I changed it to use the function that probably isn't correct. My own system had a huge number for the file-max file, so it was the wrong file.

I can look into my code and pull the function from the past commits and get the limits from another file. I am confused on wether `/proc/<pid>/limits` is what we want or `/proc/sys/fs/file-max` is what we want and I got for some reason a very high percentage.

Anyways, In a few days hopefully I will find the time to look into this, if we find the correct file for each system it sounds like an easy fix provided I can pull my function from the past commit!

**Comment by arnowelzel** [comment]
> > As a default I would set this variable to 1024 - this is the default on Linux.
> 
> This was my initial way of doing this, I had a function that was reading from a file the limits and initializing as you said to 1024 (although I was pointing to file-max which was wrong). But from the suggestion [#10192 (comment)](https://github.com/netdata/netdata/pull/10192#issuecomment-731748674) I changed it to use the function that probably isn't correct. My own system had a huge number for the file-max file, so it was the wrong file.
> 
> I can look into my code and pull the function from the past commits and get the limits from another file. I am confused on wether `/proc/<pid>/limits` is what we want or `/proc/sys/fs/file-max` is what we want and I got for some reason a very high percentage.

Just for clarification:

`/proc/sys/fs/file-max` is the system wide total of allowed open files for all processes together while `ulimit -n` or `sysconf(_SC_OPEN_MAX)` is the limit per process.

Eventhough it seems logical to check the number of open files of each single process compared to its allowed maximum, this may not be enough. Even if a process stays within its limits, the total number of open files may still be too much, just because there are many processes running trying to open files.

For example:

Let's assume an attacker starts a DDoS and causes many worker processes to be started. Each process opens 300 files, which is within the allowed limit of 1024 per process. However, when 400 of these processes got started, this will create 120000 open file descriptors which may be already above the allowed total limit of the system. In this case no alert would be triggered, since each individual process is still within its allowed limit, even when the total number of allowed file descriptors is already reached.

At least there should also be a warning, if the total number open file descriptors compared to `/proc/sys/fs/file-max` reaches a certain limit.

**Comment by ilyam8** [comment]
To be honest i think it is impossible to create default `fd usage more then N %` alarm using `apps.plugin` metrics.

There are 2 cases:
 - per system open fd limit
 - per process open fd limit

## per system open fd limit

Sum of all dimensions of `apps.files` chart == system total open fd. **We can create an alarm for total system open fd**.

## per process open fd limit

`apps.plugin` doesn't provide per process metrics, they are per groups. Group is anything a simple pattern matches. **We can't create per process alarms for open fd** because we have no per process metrics.

---

So i suggest we add a variable that is equal to `/proc/sys/fs/file-max` to the `apps.files` chart and add system total open fd percentage alarm.

**Comment by ilyam8** [comment]
On the other hand

```cmd
[ilyam@pc ~]$ cat /proc/sys/fs/file-max
9223372036854775807
```

I guess there is no much sense in having `system open fd` alarm in 2021. @Ferroin @mfundul correct me if i m wrong!

**Comment by Ancairon** [comment]
> On the other hand
> 
> ```cmd
> [ilyam@pc ~]$ cat /proc/sys/fs/file-max
> 9223372036854775807
> ```
> 
> I guess there is no much sense in having `system open fd` alarm in 2021. @Ferroin @mfundul correct me if i m wrong!

That's the thing🤣🤣 This number is huge so most people would have 0%...

**Comment by ilyam8** [comment]
Yeah, it is int64 max value 😄

---

> apps.files chart == system total open fd

That statement is wrong, it is only [FILETYPE_FILE](https://github.com/netdata/netdata/blob/47e99e48694f94fd22c91bc5267d91eb5e9f1a81/collectors/apps.plugin/apps_plugin.c#L2950-L2952) open files.

https://github.com/netdata/netdata/blob/47e99e48694f94fd22c91bc5267d91eb5e9f1a81/collectors/apps.plugin/apps_plugin.c#L253-L261

---

@mfundul suggest to use `/proc/sys/fs/file-nr` to gather both open fd anx max open fd for the sytem.

[From the docs](https://www.kernel.org/doc/Documentation/sysctl/fs.txt):

> The three values in
file-nr denote the number of allocated file handles, the number
of allocated but unused file handles, and the maximum number of
file handles. 

```cmd
[ilyam@pc ~]$ cat /proc/sys/fs/file-nr
7136	0	9223372036854775807
```

**Comment by ilyam8** [comment]
The issue was fixed in #10328.
